oauth/lib/oauth2.cpp

```diff
@@ -0,0 +1,987 @@
+        // require the client secret in PKCE flow...
+        query.addQueryItem("client_secret", d->m_clientSecret);
+    }
+
+    performAccessTokenRequest(query);
+}
+
+void SailfishOS::OAuth::OAuth2::performAccessTokenRequest(const QUrlQuery &query)
+{
+    Q_D(OAuth2);
+    const QString endpoint = (query.hasQueryItem("refresh_token") && !d->m_refreshEndpoint.isEmpty())
+                           ? d->m_refreshEndpoint : d->m_tokenEndpoint;
+    QUrl url(endpoint);
+    url.setQuery(query);
+    const QByteArray postData = url.query(QUrl::FullyEncoded).toUtf8();
+    const QUrl tokenUrl(endpoint);
+    QNetworkRequest request(tokenUrl);
+    request.setHeader(QNetworkRequest::ContentTypeHeader, "application/x-www-form-urlencoded");
+    request.setHeader(QNetworkRequest::ContentLengthHeader, postData.size());
+    if (!d->m_userAgent.isEmpty()) {
+        request.setHeader(QNetworkRequest::UserAgentHeader, d->m_userAgent);
+    }
+
+    if (!d->m_qnam) {
+        d->m_qnam = new QNetworkAccessManager(this);
+    }
+
+    QNetworkReply *accessTokenRequest = d->m_qnam->post(request, postData);
+    if (!accessTokenRequest) {
+        d->m_error = NetworkError;
+        d->m_errorMessage = QStringLiteral("Could not request access token");
+        emit errorChanged();
+        emit errorMessageChanged();
+        return;
+    }
+
+    if (!d->m_ignorableSslErrors.isEmpty()) {
+        accessTokenRequest->ignoreSslErrors(d->m_ignorableSslErrors);
+    }
+
+    connect(accessTokenRequest, &QNetworkReply::sslErrors,
+            [this, accessTokenRequest](const QList<QSslError> &errors) {
+        QStringList sslErrors;
+        for (const QSslError &error : errors) {
+            sslErrors.append(error.errorString());
+        }
+        accessTokenRequest->setProperty("sslErrors", sslErrors.join(' '));
+        emit this->sslErrors(errors);
+    });
+
+    connect(accessTokenRequest, &QNetworkReply::finished,
+            [this, accessTokenRequest]() {
+        accessTokenRequest->deleteLater();
+        const QString sslErrors = accessTokenRequest->property("sslErrors").toString();
+        if (!sslErrors.isEmpty()) {
+            d_ptr->m_error = SslError;
+            d_ptr->m_errorMessage = QStringLiteral("SSL error: %1").arg(sslErrors);
+            emit errorChanged();
+            emit errorMessageChanged();
+        } else if (accessTokenRequest->error() != QNetworkReply::NoError) {
+            const int httpCode = accessTokenRequest->attribute(QNetworkRequest::HttpStatusCodeAttribute).toInt();
+            const QByteArray replyData = accessTokenRequest->readAll();
+            d_ptr->m_error = HttpError;
+            d_ptr->m_errorMessage = QStringLiteral("HTTP error: %1: %2: %3")
+                                       .arg(httpCode)
+                                       .arg(accessTokenRequest->errorString())
+                                       .arg(QString::fromUtf8(replyData));
```

Comment:
For non-HTTP type errors, replyData would be empty. Formatting will be off a little.

Reply:
Good point, I will add a branch for that case and produce two different error messages to avoid confusion to the developer in that case.

Reply:
done